core/src/test/java/org/apache/brooklyn/enricher/stock/PercentageEnricherTest.java

```diff
@@ -0,0 +1,275 @@
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, 25d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void oneHundredPercentTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, 50d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, 100d);
+    }
+
+    @Test
+    public void negativeCurrent() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, -50d);
+        app.sensors().set(totalSensor, 100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void negativeTotal() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, -100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void bothSourceNegative() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, -50d);
+        app.sensors().set(totalSensor, -100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void totalDoubleMaxValue() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, Double.MAX_VALUE);
+        app.sensors().set(totalSensor, Double.MAX_VALUE);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, 100d);
+    }
+
+    //SETUP TESTS
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void totalNoCurrentSensor() {
+        app.enrichers().add(EnricherSpec.create(PercentageEnricher.class)
+                .configure(PercentageEnricher.SOURCE_TOTAL_SENSOR, totalSensor)
+                .configure(PercentageEnricher.TARGET_SENSOR, targetSensor));
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void totalNoTotalSensor() {
+        app.enrichers().add(EnricherSpec.create(PercentageEnricher.class)
+                .configure(PercentageEnricher.SOURCE_CURRENT_SENSOR, currentSensor)
+                .configure(PercentageEnricher.TARGET_SENSOR, targetSensor));
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void totalNoTargetSensor() {
+        app.enrichers().add(EnricherSpec.create(PercentageEnricher.class)
+                .configure(PercentageEnricher.SOURCE_CURRENT_SENSOR, currentSensor)
+                .configure(PercentageEnricher.SOURCE_TOTAL_SENSOR, totalSensor));
+    }
+
+    @Test
+    public void testDifferentProducer() {
+        EntitySpec<TestApplication> appSpec = EntitySpec.create(TestApplication.class)
+                .configure(BrooklynConfigKeys.SKIP_ON_BOX_BASE_DIR_RESOLUTION, shouldSkipOnBoxBaseDirResolution());
+
+        TestApplication producer = mgmt.getEntityManager().createEntity(appSpec);
```

Comment:
Personal preference to just create a child entity as the producer, rather than a new app - it's less code:

```
Entity producer = app.addChild(EntitySpec.create(TestEntity.class));
```